Model `Core.Int` as a class type

zygoloid opened the pull request:
Instead of treating `Core.Int` as the toolchain's builtin `IntType`, model it as a class that adapts the builtin type. This aligns us better with the intended language model, gives an associated library for `impl`s involving `Core.Int` to live within, and opens the door adding member functions to `Core.Int` if we decide that is desirable. Remarkably it also seems to make the formatted SemIR a little smaller, because a call to a generic class generates less IR than a call to a function.

zygoloid committed:
Model `Core.Int` as a class type.

zygoloid committed:
Stringify `Core.Int(N)` and `Core.UInt(N)` as `iN` and `uN`.

zygoloid committed:
Factor out code for stringifying `i32` and also use it from instnamer.

zygoloid committed:
clang-format

zygoloid commented:
Depends on #4662.

geoffromer commented on toolchain/sem_ir/type_info.h:
```diff
@@ -147,6 +147,41 @@ struct ReturnTypeInfo {
 };
 
+// Information about the type literal that corresponds to a type.
+struct TypeLiteralInfo {
```
Is this intended to eventually cover things like tuple and struct type literals? If not, should it be something like `NumericTypeLiteralInfo`?

zygoloid replied:
This is only supposed to cover numeric type literals. Renamed.

geoffromer commented on toolchain/sem_ir/stringify_type.cpp:
```diff
@@ -293,20 +320,21 @@ auto StringifyTypeExpr(const SemIR::File& sem_ir, InstId outer_inst_id)
         }
         break;
       }
       case CARBON_KIND(IntType inst): {
```
What's the role of `IntType` after this change? When is it used rather than `ClassType`?

zygoloid replied:
`IntType` is a primitive type defined by the compiler that is not a class type. `Core.Int(N)` is a class that adapts an `IntType` -- so it has the same representation as the compiler primitive type, both in SemIR and in LLVM IR, but is a class type and can have member functions and such.

geoffromer replied:
Would it be worth updating the comment on `IntType` in `typed_insts.h` to clarify this?

zygoloid replied:
Done.

geoffromer commented on toolchain/check/testdata/builtins/int/convert_checked.carbon:
```diff
@@ -1002,23 +918,19 @@ let convert_not_constant_widen: i64 = Int32ToInt64(not_constant);
 // CHECK:STDOUT:   %Int32ToUint32.type: type = fn_type @Int32ToUint32 [template]
 // CHECK:STDOUT:   %Int32ToUint32: %Int32ToUint32.type = struct_value () [template]
-// CHECK:STDOUT:   %i32: type = int_type signed, %int_32 [template]
+// CHECK:STDOUT:   %Int.2: type = class_type @Int, @Int(%int_32) [template]
```
Why isn't this named `%i32`?

zygoloid replied:
That's a bug; looks like the changes to `import_ref.cpp` in https://github.com/carbon-language/carbon-lang/commit/47285b62072a842893828f16ae36bcc765195ea9 didn't get import of package namespaces quite right. I've added a fix to this PR; let me know if you'd like it factored out to a separate PR.

geoffromer commented on toolchain/check/testdata/builtins/int/eq.carbon:
```diff
@@ -37,30 +37,28 @@ fn WrongResult(a: i32, b: i32) -> i32 = "int.eq";
 // CHECK:STDOUT:   %Bool: %Bool.type = struct_value () [template]
-// CHECK:STDOUT:   %Eq.type: type = fn_type @Eq [template]
-// CHECK:STDOUT:   %Eq: %Eq.type = struct_value () [template]
+// CHECK:STDOUT:   %Eq.type.1: type = fn_type @Eq.1 [template]
```
It'd be nice to fix these seemingly-unnecessary `.1` suffixes, but probably not a huge deal.

zygoloid replied:
Yeah. I think what's happening here is that we're importing (and then suppressing) the `Core.Int` impls from the prelude, which causes the `Core.Eq` interface to get pulled in too, and the type of the value of the interface declaration gets allocated the name `Eq.type.0`. Once we improve impl lookup to pull in fewer impls (which is on the toolchain task list), this should go away.

zygoloid committed:
Autoupdate after rebase

geoffromer commented on toolchain/check/testdata/class/adapter/adapt_copy.carbon:
```diff
@@ -105,10 +105,9 @@ fn H(a: AdaptNoncopyableIndirect) -> AdaptNoncopyableIndirect {
-// CHECK:STDOUT:   %complete_type: <witness> = complete_type_witness %i32 [template]
+// CHECK:STDOUT:   %i32: type = class_type @Int, @Int(%int_32) [template]
+// CHECK:STDOUT:   %i32.builtin: type = int_type signed, %int_32 [template]
+// CHECK:STDOUT:   %complete_type.2: <witness> = complete_type_witness %i32.builtin [template]
```
Should this be attaching to `%i32` rather than `%i32.builtin`? Among other things, the comment on `CompleteTypeWitness` indicates that it's only created for class types.

zygoloid replied:
The operand of `complete_type_witness` is the object representation of the type whose completeness it represents, so this is indicating that `AdaptCopyable` (which uses this value) is complete and its object representation is the builtin 32-bit `IntType`.

geoffromer reviewed (commented):
I'm only partway through the `testdata` changes, but those are going to take a while, so I'm sending what I have so far. Apologies if any of my comments are out of date; the PR was rebased while I was reviewing it.

zygoloid committed:
Fix build of `sieve` example.

- Remove repeated definitions of `i32` operations that were causing
  duplicate name errors.
- Build object files for the prelude and link against them when building
  `carbon_binary` rules.

zygoloid committed:
type literal -> numeric type literal

zygoloid committed:
Fix importing of namespaces corresponding to packages.

Set their parent as the current package and mark them as being closed so
that we can later recognize them as packages.

Fixes pretty-printing of an indirectly imported `Core.Int`.

zygoloid committed:
Merge branch 'trunk' into toolchain-int-is-a-class-type

zygoloid committed:
Fix up after merge.

geoffromer commented on toolchain/check/testdata/class/generic/complete_in_conversion.carbon:
```diff
@@ -31,37 +33,39 @@ fn F(a: A(0)*) {
 // CHECK:STDOUT: --- fail_derived_to_base.carbon
 // CHECK:STDOUT:
 // CHECK:STDOUT: constants {
+// CHECK:STDOUT:   %IntLiteral.type: type = fn_type @IntLiteral [template]
+// CHECK:STDOUT:   %IntLiteral: %IntLiteral.type = struct_value () [template]
+// CHECK:STDOUT:   %Int.type.1: type = fn_type @Int.1 [template]
+// CHECK:STDOUT:   %Int.1: %Int.type.1 = struct_value () [template]
```
Do I understand correctly that names like `%Int.1` and `@Int.1` refer to the local name `Int`, but `%IntLiteral` and `@IntLiteral` refer to the imported `Core.IntLiteral`, and `%Int.2` and `@Int.2` refer to the imported `Core.Int`?

zygoloid replied:
Yes.

It looks like:
- `%Int.type.1` is the type of the function `package.Int`
- `%Int.1` is the (empty) value of the function `package.Int`
- `@Int.1` is the function entity `package.Int`
- `%IntLiteral.type` is the type of the function `Core.IntLiteral`
- `%IntLiteral` is the value of the function `Core.IntLiteral`
- `@IntLiteral` is the function entity `Core.IntLiteral`
- `@Int.2` is the generic class entity `Core.Int`

I don't see a `%Int.2` in this test output, but it's probably being suppressed because it's not actually referenced, and is probably the name we assigned to the value of the generic class name `Core.Int`.

If you're asking whether the `.N` suffixes will *always* match up like this, though, the answer is no. In general the disambiguators appearing in different SemIR names are determined independently, so it's just by luck that the numbers are the same in the above example. It's not relying on a *lot* of luck though; the constants are going to be created at the same time as the entity, so there is a reason they'll often end up matching, but it's not guaranteed. But if for whatever reason we created another constant that we wanted to call `%Int`, we wouldn't leave a gap in the numbering of `@Int`s or `%Int.type`s to make the disambiguators the same.

geoffromer commented on toolchain/check/testdata/class/generic/complete_in_conversion.carbon:
```diff
@@ -70,54 +74,70 @@ fn F(a: A(0)*) {
 // CHECK:STDOUT:     import Core//prelude/...
 // CHECK:STDOUT:   }
-// CHECK:STDOUT:   %import_ref.1: %Int.type = import_ref Core//prelude/types, Int, loaded [template = constants.%Int]
+// CHECK:STDOUT:   %import_ref.1: %IntLiteral.type = import_ref Core//prelude/types/int_literal, IntLiteral, loaded [template = constants.%IntLiteral]
```
Why is the type represented as `%IntLiteral.type`, and not `Core.IntLiteral` like on line 43?

zygoloid replied:
`%IntLiteral.type` is the type of the *function* `Core.IntLiteral`. The text on line 43 is the way we pretty-print the singleton instruction ID for the `IntLiteralType` instruction.

This is made especially confusing because right now the prelude exposes a `Core.IntLiteral()` function that returns the int literal type -- per the design, this should instead be a `Core.IntLiteral` value that *is* the int literal type, but we don't actually have any working functionality in the toolchain to define that value. (`let IntLiteral:! type = MakeIntLiteralType();` doesn't work because `let :!` bindings don't work, and an alias doesn't work because we disallow aliases with non-trivial right-hand sides such as function calls.) So printing the singleton instruction ID for the int literal type as `Core.IntLiteral` doesn't accurately reflect what's currently in the prelude and is more... aspirational.

geoffromer commented on toolchain/check/testdata/function/builtin/no_prelude/import.carbon:
```diff
@@ -77,101 +77,101 @@ var arr: [i32; Core.AsIntLiteral(Core.TestAdd(Core.AsI32(1), Core.AsI32(2)))] =
 // CHECK:STDOUT:     %a.param_patt: Core.IntLiteral = value_param_pattern %a.patt, runtime_param0
-// CHECK:STDOUT:     %return.patt: %i32 = return_slot_pattern
-// CHECK:STDOUT:     %return.param_patt: %i32 = out_param_pattern %return.patt, runtime_param1
+// CHECK:STDOUT:     %return.patt: %i32.builtin = return_slot_pattern
```
Why do all these uses of `i32` refer to the builtin, rather than to `Core.Int(32)`?

zygoloid replied:
This is a no-prelude test (because it's got `/no_prelude/` in its path), so the real `Core.Int` is not available here. Instead it defines its own `Core.Int` that just returns the primitive `IntType` rather than a class that adapts it. So the type being used here is the primitive type, hence the name `%i32.builtin`.

geoffromer commented on toolchain/check/testdata/function/generic/resolve_used.carbon:
```diff
@@ -7,6 +7,7 @@
 // TIP:   bazel test //toolchain/testing:file_test --test_arg=--file_tests=toolchain/check/testdata/function/generic/resolve_used.carbon
 // TIP: To dump output, run:
 // TIP:   bazel run //toolchain/testing:file_test -- --dump_output --file_tests=toolchain/check/testdata/function/generic/resolve_used.carbon
+// CHECK:STDERR: fail_todo_call_monomorphization_error.carbon: error: integer type width of 0 is not positive [IntWidthNotPositive]
```
This diagnostic has gotten detached from its location. Is that a known issue?

zygoloid replied:
It was, though I hadn't gotten to the bottom of why it was happening. I dug into this a bit more and found it's actually a fairly shallow bug -- we just weren't passing in a source location in one key place. Fixed in 70cd03c72b9037463c8e6a1ed78f655087a09d2b.

geoffromer commented on toolchain/check/testdata/function/generic/resolve_used.carbon:
```diff
@@ -17,16 +18,14 @@ fn ErrorIfNIsZero(N:! Core.IntLiteral()) {
+  // CHECK:STDERR: fail_todo_call_monomorphization_error.carbon:[[@LINE+4]]:3: note: in `ErrorIfNIsZero(0)` used here [ResolvingSpecificHere]
   // CHECK:STDERR:   ErrorIfNIsZero(0);
   // CHECK:STDERR:   ^~~~~~~~~~~~~~
+  // CHECK:STDERR: fail_todo_call_monomorphization_error.carbon: note: in `i0` used here [ResolvingSpecificHere]
```
This also seems like it's in the wrong place.

zygoloid replied:
Also fixed in 70cd03c72b9037463c8e6a1ed78f655087a09d2b.

zygoloid committed:
Merge branch 'trunk' into toolchain-int-is-a-class-type

zygoloid committed:
Improve comment.

zygoloid committed:
Autoupdate

zygoloid committed:
Preserve the location in `require_complete_type` evaluations.

Also fix diagnostic annotation scopes being run in reverse order,
resulting in reversed context notes.

zygoloid committed:
Undo test changes made while debugging.

zygoloid committed:
Merge branch 'trunk' into toolchain-int-is-a-class-type

zygoloid committed:
Autoupdate after merge.

danakj commented on core/prelude/types/int.carbon:
```diff
@@ -2,12 +2,36 @@
+  fn Convert[self: Self]() -> Int(N) = "int.convert_checked";
+}
+
+impl forall [N:! IntLiteral()] Int(N) as ImplicitAs(IntLiteral()) {
```
Question: Why do we want to convert from an `Int(N)` to an `int_literal`? I presume this would only happen in constant evaluation? But this feels like it implies something is expecting an input of `int_literal` when it should be expecting an `int`?

zygoloid replied:
In `[T; N]` we convert `T` to `type` and `N` to `IntLiteral`. Similarly in `tuple.(N)`, we convert `N` to `IntLiteral`. So this allows things like:
```carbon
fn F(N:! i32) -> [T; N];
```

danakj replied:
Oh I see, we want to use `i32` in compile-time bindings. The input to it will have to be compile-time so it would be an `IntLiteral`, but we don't want the user to have to write that as a separate type, the `!` already covers what they need to know. Thanks.
